File: rocket-bi-web/di-core/common/domain/model/directory/Directory.ts
/*
 * @author: tvc12 - Thien Vi
 * @created: 5/30/21, 10:11 PM
 */

import { DashboardId, DDate, DirectoryId, UserId } from '../DefinedType';
import { DirectoryType, UserProfile } from '@core/common/domain/model';
import { DefaultDirectoryId } from '@/screens/directory/views/mydata/DefaultDirectoryId';
import { DirectoryMetadata } from '@core/common/domain/model/directory/directory-metadata/DirectoryMetadata';

export class Directory {
  public id: DirectoryId;
  public name: string;
  public owner?: UserProfile;
  public createdDate: DDate;
  public updatedDate: DDate;
  public ownerId: UserId;
  public parentId: DirectoryId;
  public directoryType: DirectoryType;
  public data?: DirectoryMetadata;
  public dashboardId?: DashboardId;
  public isStarred: boolean;

  constructor(
    id: DirectoryId,
    name: string,
    owner: UserProfile,
    createdDate: DDate,
    ownerId: UserId,
    parentDirectoryId: DirectoryId,
    directoryType: DirectoryType,
    updatedDate: DDate,
    data?: DirectoryMetadata,
    dashboardId?: DashboardId,
    isStar?: boolean
  ) {
    this.id = id;
    this.name = name;
    this.owner = owner;
    this.createdDate = createdDate;
    this.ownerId = ownerId;
    this.parentId = parentDirectoryId;
    this.directoryType = directoryType;
    this.dashboardId = dashboardId;
    this.data = data;
    this.updatedDate = updatedDate;
    this.isStarred = isStar ?? false;
  }

  static fromObject(obj: Directory): Directory {
    const owner = obj.owner ? UserProfile.fromObject(obj.owner) : UserProfile.unknown();
    return new Directory(
      obj.id,
      obj.name,
      owner as UserProfile,
      obj.createdDate,
      obj.ownerId,
      obj.parentId,
      obj.directoryType,
      obj.updatedDate,
      obj?.data ? DirectoryMetadata.fromObject(obj as Directory) : undefined,
      obj.dashboardId,
      obj.isStarred
    );
  }

  static default(directoryType: DirectoryType) {
    return new Directory(-100, '', UserProfile.unknown(), Date.now(), '-1', -1, directoryType, Date.now(), DirectoryMetadata.default(directoryType), -1, false);
  }

  get isUpdateDirectory() {
    return this.id >= DefaultDirectoryId.Trash;
  }
}

File: rocket-bi-server/src/main/scala/co/datainsider/bi/domain/Directory.scala
package co.datainsider.bi.domain

import co.datainsider.bi.domain.Ids.{Date, DirectoryId, UserId}
import co.datainsider.bi.domain.DirectoryType.DirectoryType
import co.datainsider.bi.domain.request.CreateDirectoryRequest
import com.fasterxml.jackson.module.scala.JsonScalaEnumeration

object Directory {
  val MyData = -1L
  val Shared = -2L

  def getSharedDirectory(orgId: Long, username: String): Directory = {
    Directory(
      orgId = orgId,
      id = Directory.Shared,
      name = "root",
      creatorId = username,
      ownerId = username,
      createdDate = System.currentTimeMillis(),
      parentId = -999
    )
  }
}

case class Directory(
    orgId: Long,
    id: DirectoryId,
    name: String,
    creatorId: UserId,
    ownerId: UserId,
    createdDate: Date,
    parentId: DirectoryId,
    isRemoved: Boolean = false,
    @JsonScalaEnumeration(classOf[DirectoryTypeRef])
    directoryType: DirectoryType = DirectoryType.Directory,
    dashboardId: Option[Long] = None,
    updatedDate: Option[Long] = None,
    data: Option[Map[String, Any]] = None
) {
  def toCreateDirRequest(parentId: DirectoryId): CreateDirectoryRequest = {
    CreateDirectoryRequest(
      name = this.name,
      parentId = parentId,
      directoryType = this.directoryType,
      dashboardId = this.dashboardId,
      isRemoved = this.isRemoved,
      data = this.data
    )
  }
}
